client/components/data/query-sites/index.jsx

```diff
@@ -7,29 +7,72 @@
 import PropTypes from 'prop-types';
 import { Component } from 'react';
 import { connect } from 'react-redux';
+import { some, forEach, isEqual, without } from 'lodash';
 
 /**
  * Internal dependencies
  */
-import { isRequestingSites, isRequestingSite } from 'state/sites/selectors';
+import { isRequestingSites, isRequestingSite, hasAllSitesList } from 'state/sites/selectors';
 import { requestSites, requestSite } from 'state/sites/actions';
+import { getPreference } from 'state/preferences/selectors';
+import { getPrimarySiteId } from 'state/selectors';
+
+const isRequestingSiteFactory = state => {
+	return siteId => isRequestingSite( state, siteId );
+};
 
 class QuerySites extends Component {
 	componentWillMount() {
-		this.request( this.props );
+		this.requestAll( this.props );
+		this.requestPrimary( this.props );
+		this.requestRecent( this.props );
+		this.requestSingle( this.props );
 	}
 
 	componentWillReceiveProps( nextProps ) {
 		if ( nextProps.siteId !== this.props.siteId ) {
-			this.request( nextProps );
+			this.requestSingle( nextProps );
+		}
+
+		if ( nextProps.primaryAndRecent && nextProps.primarySiteId !== this.props.primarySiteId ) {
+			this.requestPrimary( nextProps );
+		}
+
+		if (
+			nextProps.primaryAndRecent &&
+			! isEqual( nextProps.recentSiteIds, this.props.recentSiteIds )
+		) {
+			this.requestRecent( nextProps );
 		}
 	}
 
-	request( props ) {
+	requestAll( props ) {
 		if ( props.allSites && ! props.requestingSites ) {
 			props.requestSites();
 		}
+	}
+
+	requestPrimary( props ) {
+		if ( props.primaryAndRecent && ! props.hasAllSitesList ) {
+			const { primarySiteId, isRequestingPrimarySite } = props;
+
+			if ( primarySiteId && ! isRequestingPrimarySite ) {
+				props.requestSite( primarySiteId );
+			}
+		}
+	}
+
+	requestRecent( props ) {
```

Comment:
Could you unify the `requestRecent` and `requestPrimary` methods into one `requestPrimaryAndRecent`?

You can figure out the list of IDs to fetch by `_.union(recentList, primaryId)` and when the list changes, find the list of new sites to fetch by `_.difference(newList, oldList)`.

Note that if the list changes from `[1,2,3]` to `[1,3,2]` or `[1,3]`, you don't need to fetch anything new. It's only when new IDs appear, i.e., the list becomes `[1,2,3,4]`, then you need to fetch `[4]`. That's what `_.difference` returns.

Reply:
I don't think we need to unify them. I explicitly split them into two methods. Why do you think it would help?

Reply:
I think it could simplify the component code. And the components API triggers both requests or none of them: there is `requestPrimaryAndRecent` prop, not separate `requestRecent` and `requestPrimary`. But it's no big deal, doesn't really matter.